bigchaindb/util.py

```diff
@@ -31,6 +34,53 @@ def start(self):
             self.processes.append(proc)
 
 
+# Inspired by:
+# - http://stackoverflow.com/a/24741694/597097
+def pool(builder, size, timeout=None):
+    """Create a pool that imposes a limit on the number of stored
+    instances.
+
+    Args:
+        builder: a function to build an instance.
+        size: the size of the pool.
+
+    Returns:
+        A context manager that can be used with the ``with``
+        statement.
+    """
+
+    lock = threading.Lock()
+    local_pool = queue.Queue()
+    current_size = 0
+
+    @contextlib.contextmanager
+    def pooled():
+        nonlocal current_size
```

Comment:
> Shouldnt you retrieve current_size from the local_pool size

No, because `local_pool.size()` is just the number of instances available in the pool. We need to keep reference of the total number of instances created (that will converge to `size` after some time). I think [this test](https://github.com/bigchaindb/bigchaindb/pull/183/files#diff-5832adf374920d75d4bf48e546367f53R40) should help understanding the logic.